x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -217,62 +224,91 @@ function getDuration(items: IWaterfallItem[]) {
   return timestampEnd - timestampStart;
 }
 
+function getWaterfallItemTransactionById(
+  itemsById: IWaterfallIndex,
+  id?: IWaterfallItem['id']
+): IWaterfallItemTransaction | void {
+  if (!id) {
+    return undefined;
+  }
+
+  const item = itemsById[id];
+  if (item.docType === 'transaction') {
+    return item;
+  }
+}
+
 function createGetTransactionById(itemsById: IWaterfallIndex) {
   return (id?: IWaterfallItem['id']) => {
-    if (!id) {
-      return;
+    const waterfallItemTransaction = getWaterfallItemTransactionById(
+      itemsById,
+      id
+    );
+    if (waterfallItemTransaction) {
+      return waterfallItemTransaction.transaction;
     }
+  };
+}
 
-    const item = itemsById[id];
-    if (item.docType === 'transaction') {
-      return item.transaction;
+function createGetErrorCountByTransactionId(itemsById: IWaterfallIndex) {
+  return (id?: IWaterfallItem['id']) => {
+    const waterfallItemTransaction = getWaterfallItemTransactionById(
+      itemsById,
+      id
+    );
+    if (waterfallItemTransaction) {
+      return waterfallItemTransaction.errorCount;
     }
   };
 }
 
 export function getWaterfall(
-  hits: Array<Span | Transaction>,
-  entryTransaction: Transaction
+  hits: TraceAPIResponse['trace'],
+  errorsPerTransaction: TraceAPIResponse['errorsPerTransaction'],
+  entryTransactionId?: Transaction['transaction']['id']
```

Comment:
I'm not 100% decided on this, but I find it a bit easier to read more generic type annotations instead of the very custom ones. For instance:
```ts
  hits: Array<Span | Transaction>,
  errorsPerTransaction: StringMap<number>,
  entryTransactionId?: string
```

This is totally up for debate. Just putting out some thoughts.